contracts/src/BeefyClient.sol

```diff
@@ -390,49 +414,24 @@ contract BeefyClient {
         }
     }
 
-    /**
-     * @dev Calculate minimum number of required signatures for the current validator set.
-     *
-     * This function approximates f(x) defined below for x in [1, 21_846):
-     *
-     *  x <= 10: f(x) = x * (2/3)
-     *  x  > 10: f(x) = max(10, ceil(log2(3 * x))
-     *
-     * Research by W3F suggests that `ceil(log2(3 * x))` is a minimum number of signatures required to make an
-     * attack unfeasible. We put a further minimum bound of 10 on this value for extra security.
-     *
-     * If the session has less than 10 active validators it's definitely some sort of local testnet
-     * and we use different logic (minimum 2/3 + 1 validators must sign).
-     *
-     * One assumption is that Polkadot/Kusama will never have more than roughly 20,000 active validators in a session.
-     * As of writing this comment, Polkadot has 300 validators and Kusama has around 1000 validators,
-     * so we are well within those limits.
-     *
-     * In any case, an order of magnitude increase in validator set sizes will likely require a re-architecture
-     * of Polkadot that would make this contract obsolete well before the assumption becomes a problem.
-     *
-     * Constants generated with the help of scripts/minsigs.py
-     */
-    function minimumSignatureThreshold(uint256 validatorSetLen) internal pure returns (uint256) {
-        if (validatorSetLen <= 10) {
+    function minSignatures(uint256 validatorSetLen, uint256 signatureUseCount) internal pure returns (uint256) {
+        if (validatorSetLen <= BASE_SIGNATURE_COUNT) {
             return validatorSetLen - (validatorSetLen - 1) / 3;
-        } else if (validatorSetLen < 342) {
-            return 10;
-        } else if (validatorSetLen < 683) {
-            return 11;
-        } else if (validatorSetLen < 1366) {
-            return 12;
-        } else if (validatorSetLen < 2731) {
-            return 13;
-        } else if (validatorSetLen < 5462) {
-            return 14;
-        } else if (validatorSetLen < 10923) {
-            return 15;
-        } else if (validatorSetLen < 21846) {
-            return 16;
-        } else {
-            return 17;
         }
+
+        if (signatureUseCount == 0) {
+            return BASE_SIGNATURE_COUNT;
+        }
+
+        // log_2 rounded down is the index of the highest bit
+        uint256 logSignatureUse = Bits.highestBitSet(signatureUseCount);
+        // round up by adding 1 when there is more than 1 bit set in signatureUseCount
+        if (signatureUseCount & logSignatureUse != uint256(0)) {
+            ++logSignatureUse;
+        }
+        uint256 dynamicSignatures = 1 + 2 * logSignatureUse;
```

Comment:
This math produces signature counts that sometimes jump to a lower dynamic signature count. Is this expected? I made `minSignatures` external and ran the following test:
```js

    function testMinSignatures() public {
        // Test using signature counts 8 to 13
        assertEq(beefyClient.minSignatures(100, 8), 31);
        assertEq(beefyClient.minSignatures(100, 9), 33);
        assertEq(beefyClient.minSignatures(100, 10), 33);
        assertEq(beefyClient.minSignatures(100, 11), 33);
        assertEq(beefyClient.minSignatures(100, 12), 31); // At 12 it jumps back down to 31 again.
        assertEq(beefyClient.minSignatures(100, 13), 33);

        // Test using signature counts 70 to 75
        assertEq(beefyClient.minSignatures(100, 70), 39);
        assertEq(beefyClient.minSignatures(100, 71), 39);
        assertEq(beefyClient.minSignatures(100, 72), 37); // Again at 72
        assertEq(beefyClient.minSignatures(100, 73), 37); // And here at 73
        assertEq(beefyClient.minSignatures(100, 74), 39);
        assertEq(beefyClient.minSignatures(100, 75), 39);
    }
```
Either way, we need unit tests for this.

Reply:
https://github.com/Snowfork/snowbridge/pull/956/commits/57bfb94054b4206794dc7937ccce80efa90106f3

Reply:
Just a quick check on the math without the round-up across the whole `uint16` range.
The table below is to be read as when the signature count for a single validator is within the range in column 1 then it will choose the corresponding dynamic signature choice in column 2. In the worst case of a validator signature being used `65535` times, then the number of dynamic signatures needed is `55`.

Sig Count Values|Dynamic Sigs Choice
-|-
0 | 24
1 | 25
2 - 3 | 27
4 - 7 | 29
8 - 15 | 31
16 - 31 | 33
32 - 63 | 35
64 - 127 | 37
128 - 255 | 39
256 - 511 | 41
512 - 1023 | 43
1024 - 2047 | 45
2048 - 4095 | 47
4096 - 8191 | 49
8192 - 16383 | 51
16384 - 32767 | 53
32768 - 65535 | 55

Reply:
> M is the minimum number of signatures (24 in our code base).

My only concern is the gas cost.  Vaguely recall some benchmarks before shows gas cost doubled. So I'm wondering if **24**(compared with the previous 10) is absolutely required for the initial launch or we can make it configurable through a governance call?

Reply:
@vgeddes How did we land on 24 as the minimum number of signatures?

Reply:
Seems from https://hackmd.io/9OedC7icR5m-in_moUZ_WQ based on total DOT supply.

Reply:
### Before
| test/mocks/BeefyClientMock.sol:BeefyClientMock contract |                 |       |        |        |         |
|---------------------------------------------------------|-----------------|-------|--------|--------|---------|
| Deployment Cost                                         | Deployment Size |       |        |        |         |
| 2197297                                                 | 11304           |       |        |        |         |
| Function Name                                           | min             | avg   | median | max    | # calls |
| commitPrevRandao                                        | 763             | 17921 | 21097  | 21097  | 19      |
| createFinalBitfield                                     | 1134            | 9966  | 10597  | 10597  | 15      |
| createInitialBitfield                                   | 544             | 52080 | 52804  | 70348  | 48      |
| encodeCommitment_public                                 | 5556            | 5556  | 5556   | 5556   | 1       |
| initialize_public                                       | 2259            | 84250 | 92059  | 92059  | 23      |
| latestBeefyBlock                                        | 488             | 488   | 488    | 488    | 3       |
| minimumSignatureThreshold_public                        | 267             | 267   | 267    | 267    | 1       |
| setLatestBeefyBlock                                     | 20364           | 20364 | 20364  | 20364  | 1       |
| submitFinal                                             | 4829            | 40169 | 6704   | 112156 | 15      |
| submitInitial                                           | 14308           | 58987 | 61586  | 61853  | 20      |

### After 
| test/mocks/BeefyClientMock.sol:BeefyClientMock contract |                 |        |        |        |         |
|---------------------------------------------------------|-----------------|--------|--------|--------|---------|
| Deployment Cost                                         | Deployment Size |        |        |        |         |
| 2687425                                                 | 14246           |        |        |        |         |
| Function Name                                           | min             | avg    | median | max    | # calls |
| BASE_SIGNATURE_COUNT                                    | 269             | 269    | 269    | 269    | 1       |
| commitPrevRandao                                        | 763             | 17926  | 21097  | 21097  | 19      |
| createFinalBitfield                                     | 1153            | 23646  | 25159  | 26481  | 15      |
| createInitialBitfield                                   | 566             | 52131  | 52826  | 70370  | 50      |
| encodeCommitment_public                                 | 5578            | 5578   | 5578   | 5578   | 1       |
| initialize_public                                       | 7903            | 168210 | 183478 | 183478 | 23      |
| latestBeefyBlock                                        | 510             | 510    | 510    | 510    | 3       |
| minSignatures_public                                    | 247             | 2179   | 2225   | 2455   | 47      |
| setLatestBeefyBlock                                     | 20364           | 20364  | 20364  | 20364  | 1       |
| submitFinal                                             | 4851            | 81266  | 6726   | 239927 | 15      |
| submitInitial                                           | 39473           | 104226 | 108951 | 109218 | 20      |

Reply:
> vgeddes How did we land on 24 as the minimum number of signatures?

Hey, ah, I'm not sure. Was a minimal number decided upon in the meeting with W3F last week?

Reply:
I think so. Looks like `29`, although Alistair Stewart said `27` should be OK too. [Number of Signatures needed for Sampling Light client of BEEFY](https://hackmd.io/9OedC7icR5m-in_moUZ_WQ) cc @alistair-singh correct me if I'm wrong. 😄

Reply:
Final version of the signatureSampling is added in https://github.com/Snowfork/snowbridge/pull/956/commits/572ee397ac140d56a1d7eee2bb6dfeb93163450a. Need to regenerate test fixtures.

Reply:
Dropped minimum to 16 based on this comment about slash rate in [element](https://matrix.to/#/!xvbzTAMdhwhpLCmmLe:parity.io/$S8RBXphyc_2iqg0By85l9vKleW4Xpx-li75RhyQVGaI?via=parity.io&via=matrix.org&via=web3.foundation).
In change [e5c161c](https://github.com/Snowfork/snowbridge/pull/956/commits/e5c161c97fde55aac2e3bf8ace3a717776adf90a)